File: Python3/Hard/WrongAnswer/check-if-there-is-a-valid-parentheses-string-path_5.py
class Solution:
    def hasValidPath(self, grid: List[List[str]]) -> bool:
        
        def check(x, y):
            if x < 0 or x >= len(grid) or y < 0 or y >= len(grid[0]):
                return False
            if grid[x][y] == ' ':
                return False
            if grid[x][y] == '1':
                if grid[x][y-1] == '1' or grid[x][y-1] == '4':
                    return True
            if grid[x][y] == '2':
                if grid[x][y+1] == '2' or grid[x][y+1] == '3':
                    return True
            if grid[x][y] == '3':
                if grid[x-1][y] == '3' or grid[x-1][y] == '2':
                    return True
            if grid[x][y] == '4':
                if grid[x+1][y] == '4' or grid[x+1][y] == '1':
                    return True
            return False
        queue = [(0, 0)]
        grid[0][0] = ' '
        while queue:
            x, y = queue.pop(0)
            if x == len(grid) - 1 and y == len(grid[0]) - 1:
                return True
            if check(x+1, y):
                queue.append((x+1, y))
                grid[x+1][y] = ' '
            if check(x-1, y):
                queue.append((x-1, y))
                grid[x-1][y] = ' '
            if check(x, y+1):
                queue.append((x, y+1))
                grid[x][y+1] = ' '
            if check(x, y-1):
                queue.append((x, y-1))
                grid[x][y-1] = ' '
        return False

File: Java/Hard/WrongAnswer/check-if-there-is-a-valid-parentheses-string-path_5.java
class Solution {
    public boolean hasValidPath(char[][] grid) {
        
        int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int n = grid.length, m = grid[0].length;
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0});
        visited[0][0] = true;
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int r = cur[0], c = cur[1];
            int dir = -1;
            if (grid[r][c] == '/') {
                dir = (r + c) % 2 == 0 ? 3 : 1;
            } else if (grid[r][c] == '\\') {
                dir = (r + c) % 2 == 0 ? 1 : 3;
            } else {
                dir = 0;
            }
            for (int i = dir; i <= dir + 1; i++) {
                int nr = r + dirs[i % 4][0], nc = c + dirs[i % 4][1];
                if (nr >= 0 && nr < n && nc >= 0 && nc < m && !visited[nr][nc]) {
                    if (grid[nr][nc] == '/' && (i % 4 == 1 || i % 4 == 2)) {
                        continue;
                    }
                    if (grid[nr][nc] == '\\' && (i % 4 == 0 || i % 4 == 3)) {
                        continue;
                    }
                    if (grid[nr][nc] == '(' && (i % 4 == 3 || i % 4 == 2)) {
                        continue;
                    }
                    if (grid[nr][nc] == ')' && (i % 4 == 1 || i % 4 == 0)) {
                        continue;
                    }
                    visited[nr][nc] = true;
                    q.offer(new int[]{nr, nc});
                }
            }
        }
        return visited[n - 1][m - 1];
    }
}